src/mame/taito/taito_f3_v.cpp

```diff
@@ -366,94 +390,108 @@ TILE_GET_INFO_MEMBER(taito_f3_state::get_tile_info)
+	u8 flags = 0;
+	if (BIT(vram_tile,  8)) flags |= TILE_FLIPX;
+	if (BIT(vram_tile, 15)) flags |= TILE_FLIPY;
 
 	tileinfo.set(0,
 			vram_tile & 0xff,
-			(vram_tile >> 9) & 0x3f,
+			BIT(vram_tile, 9, 6),
 			flags);
 }
 
 TILE_GET_INFO_MEMBER(taito_f3_state::get_tile_info_pixel)
 {
-	int col_off;
-	u8 flags = 0;
-	int y_offs = (m_control_1[5] & 0x1ff);
-	if (m_flipscreen) y_offs += 0x100;
-
-	/* Colour is shared with VRAM layer */
-	if ((((tile_index & 0x1f) * 8 + y_offs) & 0x1ff) > 0xff)
-		col_off = 0x800 + ((tile_index & 0x1f) << 6) + ((tile_index & 0xfe0) >> 5);
-	else
-		col_off = ((tile_index & 0x1f) << 6) + ((tile_index & 0xfe0) >> 5);
+	/* attributes are shared with VRAM layer */
+	// convert the index:
+	// pixel: [0xxxxxxyyyyy]
+	//  text: [?yyyyyxxxxxx]
+	int y = BIT(tile_index, 0, 5);
+	int x = BIT(tile_index, 5, 6);
+	// HACK: [legacy implementation of scroll offset check for pixel palette mirroring]
+	// the pixel layer is 256px high, but uses the palette from the text layer which is twice as long
+	// so normally it only uses the first half of textram, BUT if you scroll down, you get
+	//   an alternate version of the pixel layer which gets its palette data from the second half of textram.
+	// we simulate this using a hack, checking scroll offset to determine which version of the pixel layer is visible.
+	// this means we SHOULD dirty parts of the pixel layer, if the scroll or flipscreen changes.. but we don't.
+	// (really we should just apply the palette during rendering instead of this ?)
+	int y_offs = y * 8 + m_control_1[5];
+	if (m_flipscreen)
+		y_offs += 0x100; // this could just as easily be ^= 0x100 or -= 0x100
+	if ((y_offs & 0x1ff) >= 256)
+		y += 32;
 
-	const u16 vram_tile = (m_textram[col_off] & 0xffff);
+	const u16 vram_tile = m_textram[y << 6 | x];
 
-	if (vram_tile & 0x0100) flags |= TILE_FLIPX;
-	if (vram_tile & 0x8000) flags |= TILE_FLIPY;
+	const int tile = tile_index;
+	const u8 palette = BIT(vram_tile, 9, 6);
+	u8 flags = 0;
+	if (BIT(vram_tile, 8)) flags |= TILE_FLIPX;
+	if (BIT(vram_tile, 15)) flags |= TILE_FLIPY;
 
-	tileinfo.set(1,
-			tile_index,
-			(vram_tile >> 9) & 0x3f,
-			flags);
+	tileinfo.set(1, tile, palette, flags);
 }
 
 /******************************************************************************/
 
 void taito_f3_state::screen_vblank(int state)
 {
-	// rising edge
-	if (state)
-	{
-		if (m_sprite_lag==2)
-		{
-			if (machine().video().skip_this_frame() == 0)
-			{
-				get_sprite_info(m_spriteram16_buffered.get());
-			}
-			memcpy(m_spriteram16_buffered.get(), m_spriteram.target(), 0x10000);
-		}
-		else if (m_sprite_lag == 1)
-		{
-			if (machine().video().skip_this_frame() == 0)
-			{
-				get_sprite_info(m_spriteram.target());
-			}
-		}
+	if (state) {
+		//get_sprite_info();
+	}
+}
+
+void taito_f3_state::set_extend(bool state) {
+	m_extend = state;
+	// TODO: we need to free these if this is called multiple times
+	// for (int i=0; i<8; i++) {
+	// 	delete m_tilemap[i];
+	// }
```

Comment:
Can you rename this to `create_tilemaps` or something to make it clearer what it does, and call the parameter `extend`?

Also, you really need to restructure the code so this *can’t* be called more than once.  Tilemaps are one of those things with session lifetime – they need to be created at start time, they get cleaned up automatically on stop, and they automatically participate in save states.

Reply:
immediate ask done, though tilemaps in f3 might need rethinking in general at some point.  we like to construct them for the debug viewer, but then have to just read the pixmap to apply line effects properly, and then they participate in additive blending.